processing/src/main/java/io/druid/segment/DimensionIndexer.java

```diff
@@ -0,0 +1,323 @@
+ * under the License.
+ */
+
+package io.druid.segment;
+
+import com.metamx.collections.bitmap.BitmapFactory;
+import com.metamx.collections.bitmap.MutableBitmap;
+import io.druid.query.dimension.DimensionSpec;
+import io.druid.query.filter.DruidPredicateFactory;
+import io.druid.query.filter.ValueMatcher;
+import io.druid.segment.data.Indexed;
+import io.druid.segment.incremental.IncrementalIndex;
+import io.druid.segment.incremental.IncrementalIndexStorageAdapter;
+
+/**
+ * Processing related interface
+ *
+ * A DimensionIndexer is a per-dimension stateful object that encapsulates type-specific operations and data structures
+ * used during the in-memory ingestion process (i.e., work done by IncrementalIndex).
+ *
+ * Ingested row values are passed to a DimensionIndexer, which will update its internal data structures such as
+ * a value->ID dictionary as row values are seen.
+ *
+ * The DimensionIndexer is also responsible for implementing various value lookup operations,
+ * such as conversion between an encoded value and its full representation. It maintains knowledge of the
+ * mappings between encoded values and actual values.
+ *
+ * NOTE:
+ * When encoding is present, there are two relevant orderings for the encoded values.
+ *
+ * 1.) Ordering based on encoded value's order of ingestion
+ * 2.) Ordering based on converted actual value
+ *
+ * Suppose we have a new String dimension DimA, which sees the values "Hello", "World", and "Apple", in that order.
+ * This would correspond to dictionary encodings of "Hello"=0, "World"=1, and "Apple"=2, by the order
+ * in which these values were first seen during ingestion.
+ *
+ * However, some use cases require the encodings to be sorted by their associated actual values.
+ * In this example, that ordering would be "Apple"=0, "Hello"=1, "World"=2.
+ *
+ * The first ordering will be referred to as "Unsorted" in the documentation for this interface, and
+ * the second ordering will be referred to as "Sorted".
+ *
+ * @param <EncodedType> class of the encoded values
+ * @param <ActualType> class of the actual values
+ *
+ */
+public interface DimensionIndexer<EncodedType extends Comparable<EncodedType>, ActualType extends Comparable<ActualType>>
+{
+  /**
+   * Given a single row value, update any internal data structures with the ingested values and
+   * return the row values as an array to be used as a TimeAndDims key.
+   *
+   * For example, the dictionary-encoded String-type column will return an int[] containing a dictionary ID.
+   *
+   * The value within the returned array should be encoded if applicable, i.e. as instances of EncodedType.
+   *
+   * NOTE: This function can change the internal state of the DimensionIndexer.
+   *
+   * @param dimValues Single row val to process
+   *
+   * @return An array containing an encoded representation of the input row value.
+   */
+  public Object processSingleRowValToIndexKey(Object dimValues);
```

Comment:
hmm, would it be workable to make the return type `EncodedArrayType` rather than `Object`?

So the specialization for String handler would be EncodedArray = int[], Encoded = Integer, Actual = String